src/main/scala/io/iohk/ethereum/blockchain/sync/NewRegularSync.scala

```diff
@@ -0,0 +1,311 @@
+    * If the execution of any block were to fail, newBlocks only contains the NewBlock msgs for all the blocks executed before it,
+    * and only the blocks successfully executed are inserted into the blockchain.
+    *
+    * @param blocks to execute
+    * @param newBlocks which, after adding the corresponding NewBlock msg for blocks, will be broadcasted
+    * @return list of NewBlocks to broadcast (one per block successfully executed) and an error if one happened during execution
+    */
+  @tailrec
+  private def processBlocks(blocks: Seq[Block], newBlocks: Seq[NewBlock] = Nil)
+  : (Seq[NewBlock], Option[BlockImportFailure]) = blocks match {
+    case Nil =>
+      newBlocks -> None
+
+    case Seq(block, otherBlocks@_*) =>
+      ledger.importBlock(block) match {
+        case BlockImported(_, td, topOfChain) =>
+          if (topOfChain) {
+            //FIXME: what about a deeper than 1 level chain re-org? Try to return a switched branch in `BlockImported`
+            pendingTransactionsManager ! PendingTransactionsManager.RemoveTransactions(block.body.transactionList)
+            ommersPool ! new RemoveOmmers((block.header +: block.body.uncleNodesList).toList)
+          }
+
+          // FIXME: again, the switched branch...
+          val updatedNewBlocks = if (topOfChain) newBlocks :+ NewBlock(block, td) else newBlocks
+
+          processBlocks(otherBlocks, newBlocks :+ NewBlock(block, td))
+
+        case error: BlockImportFailure =>
+          newBlocks -> Option(error)
+      }
+  }
+
+  private def scheduleResume() = {
+    headersQueue = Nil
+    scheduler.scheduleOnce(checkForNewBlockInterval, self, ResumeRegularSync)
+  }
+
+  private def resumeWithDifferentPeer(currentPeer: Peer, reason: String = "error in response") = {
+    blacklist(currentPeer.id, blacklistDuration, reason)
+    headersQueue = Nil
+    context.self ! ResumeRegularSync
+  }
+
+  private def bestPeer: Option[Peer] = {
+    val peersToUse = peersToDownloadFrom
+      .collect {
+        case (ref, PeerInfo(_, totalDifficulty, true, _)) => (ref, totalDifficulty)
+      }
+
+    if (peersToUse.nonEmpty) Some(peersToUse.maxBy { case (_, td) => td }._1)
+    else None
+  }
+
+}
+
+object NewRegularSync {
+  // scalastyle:off parameter.number
+  def props(appStateStorage: AppStateStorage, blockchain: Blockchain, validators: Validators,
+      etcPeerManager: ActorRef, peerEventBus: ActorRef, ommersPool: ActorRef, pendingTransactionsManager: ActorRef, ledger: Ledger,
+      syncConfig: SyncConfig, scheduler: Scheduler): Props =
+    Props(new NewRegularSync(appStateStorage, blockchain, validators, etcPeerManager, peerEventBus, ommersPool, pendingTransactionsManager,
+      ledger, syncConfig, scheduler))
+
+  private case object ResumeRegularSync
+  private case class ResolveBranch(peer: ActorRef)
+  private case object PrintStatus
+
+  case object Start
+  case class MinedBlock(block: Block)
+}
```

Comment:
is it the original RegularSync with your changes? why do we need a copy (and not modify the existing `RegularSync`)? Can you tell which areas in this file has been changed?